urld2fb1d2b05_thnaeff_Util_tgz-pJ8-ch_thn_util_tests_TreeTestJ8/src/ch/thn/util/file/PathWatcher.java

```diff
@@ -0,0 +1,594 @@
+	 * @param allParents If set to <code>true</code>, all parent directories are
+	 * registered too
+	 * @return
+	 */
+	public boolean registerPath(String path, boolean allChildren, boolean allParents) {
+		return registerPath(Paths.get(path), allChildren, allParents);
+	}
+
+	/**
+	 * Adds a new path to the list of watched paths.
+	 * 
+	 * @param path
+	 * @return
+	 */
+	public boolean registerPath(String path) {
+		return registerPath(path, false, false);
+	}
+
+	/**
+	 * Adds a new path to the list of watched paths.
+	 * 
+	 * @param path
+	 * @return
+	 */
+	public boolean registerPath(Path path) {
+		return registerPath(path, false, false);
+	}
+
+	/**
+	 * Registers the given path for all the events
+	 * 
+	 * @param dir
+	 * @param allChildren
+	 * @return
+	 * @throws IOException
+	 */
+	private boolean register(Path dir, boolean allChildren) {
+
+		WatchKey key = null;
+
+		try {
+			if (usePolling) {
+				key = ((PollingWatchService)watcher).register(dir);
+			} else {
+				key = dir.register(watcher,
+						StandardWatchEventKinds.ENTRY_CREATE,
+						StandardWatchEventKinds.ENTRY_DELETE,
+						StandardWatchEventKinds.ENTRY_MODIFY);
+			}
+		} catch (Exception e) {
+			throw new PathWatcherError("Failed to register path " + dir, e);
+		}
+
+		//If its the same one it will just be updated
+		keys.put(key, dir);
+		this.allChildren.put(key, allChildren);
+
+		fireNewPathWatched(dir);
+
+		return true;
+	}
+
+	/**
+	 * Walks through the file tree and registers all child paths
+	 * 
+	 * @param path
+	 * @throws IOException
+	 */
+	private void registerAllChildren(Path path) {
+
+		try {
+			Files.walkFileTree(path, new SimpleFileVisitor<Path>() {
+				@Override
+				public FileVisitResult preVisitDirectory(Path dir, BasicFileAttributes attrs) {
+					try {
+						register(dir, true);
+					} catch (Exception e) {
+						logger.warn("Failed to recursively register child path " + dir, e);
+					}
+					return FileVisitResult.CONTINUE;
+				}
+			});
+		} catch (IOException e) {
+			logger.warn("Failed to recursively register path and children of " + path, e);
+		}
+
+	}
+
+	/**
+	 * Walks upwards through the file tree and registers all parent paths
+	 * 
+	 * @param path
+	 * @throws IOException
+	 */
+	private void registerAllParents(Path path) {
+		Path p = path;
+
+		while (p != null) {
+			try {
+				register(p, false);
+			} catch (Exception e) {
+				logger.warn("Failed to recursively register parent path " + path, e);
+			}
+			p = p.getParent();
+		}
+
+	}
+
+
+	@SuppressWarnings("unchecked")
+	static <T> WatchEvent<T> cast(WatchEvent<?> event) {
+		return (WatchEvent<T>)event;
+	}
+
+
+	@Override
+	public void run() {
+		runStart();
+
+		while (!isStopRequested()) {
+			runPause(false);
+
+			//Check if just stop requested or pausing exited because stop has been requested
+			if (isStopRequested()) {
+				break;
+			}
+
+			WatchKey key = null;
+
+			try {
+				key = watcher.take();
+			} catch (InterruptedException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)